typed-racket-lib/typed-racket/rep/rep-utils.rkt

```diff
@@ -242,141 +323,156 @@
         [(provides ...)
          (cond
-           [(attribute no-provide?) #'()]
+           [(attribute no-provide?-kw) #'()]
            [else
-            #'((provide var.mexpdr var.pred flds.accessors ...)
+            #'((provide var.match-expander var.predicate flds.accessors ...)
                (provide/cond-contract (var.constructor constructor-contract)))])]
-        [(extra-defs ...) (if (attribute extras) #'extras #'())])
+        [(extra-defs ...) (if (attribute extras) #'extras #'())]
+        [struct-def #'(struct var.name parent ... (flds.ids ...)
+                        maybe-transparent ...
+                        #:constructor-name constructor-name
+                        #:property prop:uid uid-id
+                        #:property prop:mask rep-mask-body
+                        #:methods gen:Rep
+                        [Rep-name-def
+                         Rep-values-def
+                         Rep-constructor-def
+                         Rep-variances-def
+                         Rep-free-vars-def
+                         Rep-free-idxs-def
+                         Rep-for-each-def
+                         Rep-fmap-def]
+                        extra-defs ...)])
        ;; - - - - - - - - - - - - - - -
        ;; macro output
        ;; - - - - - - - - - - - - - - -
-       #'(begin
-           (struct var.name parent (flds.ids ...) #:transparent
-             #:constructor-name
-             var.raw-constructor
-             #:property prop:Rep-name (quote var.name)
-             #:property prop:constructor-fun
-             (λ (flds.ids ...) (var.constructor flds.ids ...))
-             #:property prop:values-fun
-             values-def
-             #:property prop:walk-fun
-             walk-def
-             #:property prop:fold-fun
-             fold-def
-             maybe-top-type-spec ...
-             maybe-structural ...
-             maybe-needs-resolving ...
-             extra-defs ...)
-           constructor-def
-           mexpdr-def
-           provides ...))]))
+       (cond
+         [(attribute singleton)
+          (syntax/loc stx
+            (begin
+              (define uid-id (next-uid!))
+              (define singleton
+                (let ()
+                  struct-def
+                  (var.constructor)))
+              (declare-predefined-type! singleton)
+              (define (var.predicate x) (eq? x singleton))
+              (define-match-expander var.match-expander
+                (λ (s)
+                  (syntax-parse s
+                    [(_) (syntax/loc s (? var.predicate))])))
+              (provide singleton var.predicate var.match-expander
+                       uid-id)))]
+         [else
+          (syntax/loc stx
+            (begin
+              (define uid-id (next-uid!))
+              struct-def
+              constructor-def
+              mexpdr-def
+              provides ...
+              (provide uid-id)))]))]))
```

Comment:
switches are building vectors that store the appropriate λ's in their corresponding field. The field index is bound to uid-id, so having it in scope makes it easy for the macro to simply expand into a reference to that id.